### PR TITLE
Faster

### PR DESCRIPTION
**Pull request by bstuder**
* Move Vector to Unboxed Version
* Fix Matrix underlying type do Double

### CODE REVIEW

File: src/Matrix.hs [hunk 1]
```diff
@@ -30,8 +32,8 @@ module Matrix
 
 import Data.Maybe (fromMaybe)
 import qualified Data.List as DL (transpose)
-import qualified Data.Vector as DVB (Vector(..))
-import qualified Data.Vector.Unboxed as DV (Vector(..), (!), and, backpermute, fromList, generate, length, map, maximum, minimum, toList, zipWith, empty)
+import qualified Data.Vector as DVB (Vector(..), toList)
+import qualified Data.Vector.Unboxed as DV (Vector(..), (++), (!), and, backpermute, fromList, generate, length, map, maximum, minimum, toList, zipWith, empty, take, concat)
```
Comment (Ostaijen) [inline src/Matrix.hs]: Hmmm... `Data.Vector as DVB` and `Data.Vector.Unboxed as DV` seems a lazzy and utterly ugly hack to make things work without proper naming...

File: src/Matrix.hs [hunk 2]
```diff
@@ -2,26 +2,28 @@
 
 module Matrix
 (
-    Matrix,
+    Matrix(..),
```
Comment (Ostaijen) [inline src/Matrix.hs]: Alphabetical order ! 😢

File: src/Matrix.hs [hunk 3]
```diff
@@ -106,39 +109,38 @@ fromLayers (x:y:xs) generator = Matrix y x randomVector : fromLayers (y:xs) seco
...
-    lowerBound = maybe (minimum matrix) id maybeLowerBound
+normalize :: (DV.Unbox a, RealFloat a) => Maybe a -> Maybe a -> Matrix a -> Matrix a
+normalize maybeUpperBound maybeLowerBound matrix = Matrix.map transform matrix
+  where transform  = if upperBound == lowerBound
+                       then const 0
+                       else \x -> (2 * x - upperBound - lowerBound) / (upperBound - lowerBound)
+        upperBound = fromMaybe (maximum matrix) maybeUpperBound
+        lowerBound = fromMaybe (minimum matrix) maybeLowerBound
 
 infixr 8 !
```
Comment (Ostaijen) [inline src/Matrix.hs]: Should be closer to the begining of the file ?

File: src/Matrix.hs (hunk 1 above)
Comment (bstuder) [reply src/Matrix.hs]: A bit, but it also avoid having a huge diff with `DV` -> `DVU`, also the Unboxed is now our main/only type of vector.

File: src/Matrix.hs (hunk 2 above)
Comment (bstuder) [reply src/Matrix.hs]: I'll allow it :p

File: src/Matrix.hs (hunk 3 above)
Comment (bstuder) [reply src/Matrix.hs]: :+1:

**Pull request merged**